### DISCUSSION TITLE
CMake projects on Windows do not copy DLLs to output folder.

### DISCUSSION THREAD

**Discussion by fishnet37222** [microsoft/vcpkg#19408] [Q&A]
**Describe the bug**
When using vcpkg to pull in dependencies as DLLs, the resulting DLLs are not copied into the output folder.

**Environment**
- OS: [Windows]
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29910 for x64

**To Reproduce**
1. Open the repository at [https://github.com/fishnet37222/HelloWorld.git](url) in either Visual Studio or Visual Studio Code
2. Build the project.
3. Inspect the contents of the output folder.

**Expected behavior**
The DLLs generated by the libraries pulled in by vcpkg should be copied to the output folder.

**Failure logs** 
No failure logs generated.

**Additional context**
The sample project I created references both wxWidgets and Boost-filesystem.
See [https://github.com/fishnet37222/vcpkg-test.git](url) for a regular Visual Studio C++ project using vcpkg in which the issue does not occur.

**Comment by NancyLi1013** [comment]
Hi @fishnet37222 

Thanks for posting this issue.

It seems that the link you provided above are wrong. Since I clicked them and both of them are directed to https://github.com/microsoft/vcpkg/issues/url.

Could you please help provide your project sample and attach your result here?

**Comment by JackBoosY** [comment]
Please see feature `X_VCPKG_APPINSTALL_DEPS_INSTALL `.
Related: https://github.com/microsoft/vcpkg/pull/14243

**Comment by fishnet37222** [comment]
I've attached the two projects as requested.

[HelloWorld.zip](https://github.com/microsoft/vcpkg/files/6099756/HelloWorld.zip)
[vcpkg-test.zip](https://github.com/microsoft/vcpkg/files/6099757/vcpkg-test.zip)

**Comment by NancyLi1013** [comment]
@fishnet37222 

Have you tried the above suggestions from @JackBoosY ?

**Comment by fishnet37222** [comment]
Is there any documentation on how to use what @JackBoosY was referring to?

**Comment by NancyLi1013** [comment]
Currently there is no any documentation about this feature. Since it is experimental.

I think you may set `X_VCPKG_APPLOCAL_DEPS_INSTALL ` as `ON ` to rebuild vcpkg and verify if this can work.

`option(X_VCPKG_APPLOCAL_DEPS_INSTALL "(experimental) Automatically copy dependencies into the install target directory for executables." OFF)`
https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L49

**Comment by fishnet37222** [comment]
I tried setting that property to `ON` in my `CMakeLists.txt` file, but the DLLs were still not copied to the output folder.  I also noticed that when I create a console application (the `add_executable` command does not include `WIN32`), the DLLs do get copied to the output folder.

**Comment by pratikpc** [comment]
@NancyLi1013 do you know if X_VCPKG_APPINSTALL_DEPS_INSTALL copies only VCPKG dependencies or if it will also copy my shared library's DLLs?

Say if I have a Windows project with 
1. Boost Asio
2. My Shared Library which generates a DLL

Would X_VCPKG_APPINSTALL_DEPS_INSTALL only copy Boost Asio or would it also copy My Shared Library's DLLs?

**Comment by JackBoosY** [comment]
@pratikpc  Yes, it should do that.

**Comment by fishnet37222** [comment]
I finally found out why the DLLs are not being copied to the binary output folder in CMake projects.  I took the time to debug the "applocal.ps1" file in VSCode and discovered that the DLL names it's looking for are not the same as the ones that were actually generated.  In my case, the script is looking for a DLL named "wxmsw315u_core_vc_custom.dll" when the file that actually exists is named "wxmsw315ud_core_vc_custom.dll".  It seems like CMake is only finding the optimized variants of the DLLs because those are the only ones I see listed in my project's "CMakeCache.txt" file.

When I tried manually copying over the optimized DLLs into the binary output folder, I still couldn't launch the executable.  I got the error shown below.
![image](https://user-images.githubusercontent.com/1977106/138764790-7cdab979-7364-429c-abac-f44d06957082.png)

**Reply by mathisloge** [reply]
I recommend that you either use [DependencyWalker](https://github.com/lucasg/Dependencies) to quickly find out which dlls aren't correct or missing or in a more complex scenario loader snaps.
